### PR TITLE
export model on every checkpoint + log printing update

### PR DESCRIPTION
**Pull request by brandondutra**
--logs now only print throughput one
--on every checkpoint, an exported model is made. Last n models are kept
as backups.

### CODE REVIEW

File: src/prediction/_model_export_manager.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+import os
+import collections
+import time
+import tensorflow as tf
+from tensorflow.python.lib.io import file_io
+import tensorfx as tfx
+
+class ModelExportManager(object):
+  def __init__(self, intermediate_model_folder, final_model_folder, exports_to_keep):
+    """Exports last model into final location, keeps last few trained models.
```
Comment (nikhilk) [inline src/prediction/_model_export_manager.py:22]: We are already writing out checkpoints along the way, and its actually straightforward to go from checkpoint to a model if needed (I wish a saved model used the checkpoint file as-is ... maybe there is a way), so this work of loading each checkpoint to save a saved model is quite wasteful (once you factor in larger models and reads/writes to GCS). So I am concerned with this change at a broader level.

Additionally the pattern of keeping N changes is problematic on GCS, since they require enumeration/copies. Something other than the training process should be in charge of deletes based on size/age/quality.

In fact what we need with multiple checkpoints/exports along the way is some place to store the eval metric achieved with that checkpoint, so a user can decide which checkpoint to use easily.

So I think what we really need is just an argument to expose the # of checkpoints to keep, and default it to 0, meaning keep-all, that is passed in into the saver construction.

File: src/training/_hooks.py [hunk 2]
```diff
@@ -43,9 +43,9 @@ def after_run(self, context, values):
 
 
 class LogSessionHook(tf.train.SessionRunHook):
-  """Logs the session loop by outputting steps, and throughput into logs.
+  """Writes training throughput stats into the logs.
   """
-  _MESSAGE_FORMAT = 'Run: %.2f sec; Steps: %d; Duration: %d sec; Throughput: %.1f instances/sec'
+  _MESSAGE_FORMAT = 'Training throughput: Step %d time: %.2f; Duration: %d sec; Lifetime throughput: %.1f instances/sec'
```
Comment (nikhilk) [inline src/training/_hooks.py:48]: Not sure I understand the change here.

Lets keep logs short, so they don't wrap on a normal console width.

1. The first number is not a throughput -- it is time taken to run N steps between successive logs. So it should probably be left at Run.
2. Lifetime is implied.
3. Steps is plural everywhere.

File: src/training/_hooks.py [hunk 3]
```diff
@@ -73,13 +73,17 @@ def after_run(self, context, values):
...
 class LogTrainingHook(tf.train.SessionRunHook):
-  """Logs the training job by logging progress as well as producing summary events.
+  """Writes training loss into the logs, and writes summary events.
+
+  Two summary events are written
+  1) training loss of the current training step
+  2) running training throughput
   """
-  _MESSAGE_FORMAT = 'Global steps: %d; Duration: %d sec; Throughput: %.1f instances/sec; Loss: %.3f'
+  _MESSAGE_FORMAT = 'Global step %d: Training loss: %.3f'
```
Comment (nikhilk) [inline src/training/_hooks.py:86]: Similarly, not sure I understand this.

1. Why remove the throughput? And duration of the job?
2. Steps is generally plural everywhere in APIs.

File: src/training/_hooks.py [hunk 4]
```diff
@@ -119,7 +123,7 @@ def after_run(self, context, values):
       throughput = self._global_steps_completed * float(self._batch_size) / float(duration)
 
       logging.info(LogTrainingHook._MESSAGE_FORMAT,
-                   self._global_steps_completed, duration, throughput, loss_value)
+                   self._global_steps_completed, loss_value)
```
Comment (nikhilk) [inline src/training/_hooks.py:126]: General comment on the log change based on the description of the CL ... this looks like logging the same thing multiple times, which I think you are addressing, but it isn't once you factor distributed training. Each worker might have different throughput values.

Maybe we need to refactor differently -- a worker logger and a master logger -- but I haven't fully thought through that. In fact, we might be able to achieve this now with just using LogTrainingHook and not LogSessionHook in the case of a single node training job.

**Review by nikhilk — CHANGES_REQUESTED** [review]

File: src/training/_hooks.py (hunk 4 above)
Comment (brandondutra) [reply src/training/_hooks.py:126]: oh, interesting point. As a smaller step, what if we prepend the word 'master' to LogTrainingHook?

File: src/training/_hooks.py (hunk 2 above)
Comment (brandondutra) [reply src/training/_hooks.py:48]: I don't think run_time is the time between log events. Before_run and after_run get called on each training step, right? So in after_run(), run_time = time of running 1 training step, not N steps.

That's why I was confused by which values cover just the last step (run, loss), the last N steps between successive logs (nothing?), and all steps (duration, throughput)

**Review by brandondutra — COMMENTED** [review]
If you agree the log names could be improved, I'll update this PR to just be about logs (I'll remove the export stuff). This was really more about me learning/messing with your framework.  

I have not seen how easy/hard it is to build a serving graph from a checkpoint. This is something I want to explore later. I like the idea of saving the checkpoint evaluations to a file someplace, and I can do this in a new PR.

**Comment by brandondutra** [comment]
Per conversation, going to close this PR and open new ones for the eval/checkpoint and less hooks for single node training.
